tripal_chado/src/Form/ChadoCustomTableForm.php

```diff
@@ -0,0 +1,331 @@
+      '#type' => 'item',
+      '#markup' => "Example library_stock table: <pre>[
+  'table' => 'library_stock',
+  'fields' => [
+    'library_stock_id' => [
+      'type' => 'serial',
+      'not null' => TRUE,
+    ],
+    'library_id' => [
+      'type' => 'int',
+      'not null' => TRUE,
+    ],
+    'stock_id' => [
+      'type' => 'int',
+      'not null' => TRUE,
+    ]
+  ],
+  'primary key' => [
+    'library_stock_id'
+  ],
+  'unique keys' => [
+    'library_stock_c1' => [
+      'library_id',
+      'stock_id'
+    ]
+  ],
+  'foreign keys' => [
+    'library' => [
+      'table' => 'library',
+      'columns' => [
+        'library_id' => 'library_id'
+      ],
+    ],
+    'stock' => [
+      'table' => 'stock',
+      'columns' => [
+        'stock_id' => 'stock_id'
+      ]
+    ]
+  ]
+]</pre>",
+    ];
+  
+    if ($action == 'Add') {
+      $form['force_drop'] = [
+        '#type' => 'value',
+        '#value' => $default_force_drop,
+      ];
+    }
+    else {
+      $form['force_drop'] = [
+        '#type' => 'checkbox',
+        '#title' => t('Re-create table'),
+        '#description' => t('Check this box if your table already exists and you would like to drop it and recreate it.'),
+        '#default_value' => $default_force_drop,
+      ];
+    }
+    $form['schema'] = [
+      '#type' => 'textarea',
+      '#title' => t('Schema Array'),
+      '#description' => t('Please enter the ' . Link::fromTextAndUrl('Drupal Schema API', Url::fromUri('https://api.drupal.org/api/drupal/includes!database!schema.inc/group/schemaapi/7', ['attributes' => ['target' => '_blank']]))->toString() . ' compatible array that defines the table.'),
+      '#required' => FALSE,
+      '#default_value' => $default_schema,
+      '#rows' => 25,
+    ];
+  
+    if ($action == 'Edit') {
+      $value = 'Save';
+    }
+    if ($action == 'Add') {
+      $value = 'Add';
+    }
+  
+    $form['submit'] = [
+      '#type' => 'submit',
+      '#value' => t($value),
+      '#executes_submit_callback' => TRUE,
+    ];
+    
+    $form['cancel'] = [
+      '#type' => 'markup',
+      '#markup' => Link::fromTextAndUrl('Cancel', Url::fromUserInput('/admin/tripal/storage/chado/custom_tables'))->toString(),
+    ];
+  
+    
+    return $form;
+  }
+  
+  /**
+   * Implements hook_validate().
+   * Validate the Create/Edit custom table form.
+   *
+   */
+   //function tripal_custom_tables_form_validate($form, &$form_state) {
+  public function validateForm(array &$form, FormStateInterface $form_state) { 
+    $values = $form_state->getValues();   
+    $action = $values['action'];
+    $table_id = $values['table_id'];
+    $schema = $values['schema'];
+    $force_drop = $values['force_drop'];
+  
+    if (!$schema) {
+      $form_state->setErrorByName($values['schema'], t('Schema array field is required.'));
+    }
+  
+    // make sure the array is valid
+    $schema_array = [];
+    if ($schema) {
+      // $success = preg_match('/^\s*array/', $schema);

```

Comment:
I believe this was removed because it was a quick check using old PHP style arrays.  Since we now allow the new style this shouldn't be needed. I will take out the commented line and push.